### PR TITLE
fix(sdk): keep a baseURL path prefix instead of discarding it

### LINKED ISSUE #428: feat(sdk): support a base path in baseURL — deployments behind a path prefix or BFF
**Area:** sdk — gap (blocks a whole deployment shape) · found while building a first-party web console against WaveHouse

**Expected:** point the SDK at a WaveHouse reachable under a **path prefix** rather than at the root of an origin — an app server, BFF, or ingress exposing it at e.g. `https://app.example.com/api/warehouse/…` — by including that prefix in `baseURL`.

**Actual:** the prefix is silently discarded. `buildURL` (`clients/ts/src/http.ts:112-120`) resolves with `new URL(path, base)`, and every SDK request path is **absolute** — `/v1/query`, `/v1/admin/pipes` (`pipes.ts:64`), `/v1/admin/policy` (`policy.ts:17`). Per the URL spec an absolute path replaces the base's path entirely:

```js
new URL('/v1/query', 'https://app.example.com/api/warehouse/')
// → 'https://app.example.com/v1/query'   ← the /api/warehouse prefix is gone
```

The streaming transport has the same construction — `new URL("/v1/stream", this._opts.baseURL)` (`stream/sse.ts:56`) — so it collapses identically.

The failure mode is the bad kind: **no error**. The client looks correctly configured, and every request quietly goes somewhere else. There's no workaround from outside the SDK, because `baseURL` is the only path input and it can't survive.

**Why it matters:** serving WaveHouse under a path prefix is an ordinary deployment shape — one origin fronting several services, a BFF adding auth or rate limiting ahead of the data API, a Next.js/Remix/Rails route proxying through, or a path-routed ingress. Every one of them is currently pushed off the SDK and onto hand-rolled `fetch`, which is the code the SDK exists to delete. It also means such deployments lose the query builder, typed errors, retry/backoff, and the streaming controller, not just the URL handling.

**Scope:** either

- resolve request paths **relative** to `baseURL` — drop the leading `/` from SDK paths and always normalize `base` to end in `/`; or
- add an explicit `basePath?: string` to `ClientConfig`, prepended ahead of the `/v1/...` path.

The first is less API surface. The second is more explicit and can't surprise anyone depending on today's resolution. Either way `stream/sse.ts` needs the same treatment — it's easy to fix `http.ts` alone and leave streaming broken — and a test pinning a prefixed base (covering both transports) would keep it fixed.

Related: #269 — the sibling gap in the same config surface (no custom headers / `fetch` override). A deployment behind a proxy usually needs both.

---
_Found while building a first-party web console against WaveHouse, 2026-08-05._

### PR DESCRIPTION
**Pull request by EricAndrechek**
Closes #428.

## The bug

Pointing the SDK at a WaveHouse served under a path prefix silently dropped the prefix:

```js
createClient({ baseURL: 'https://app.example.com/api/warehouse' })
// every request went to https://app.example.com/v1/query
```

Both transports resolved **absolute** request paths against the base — `new URL('/v1/query', base)` in `http.ts`, `new URL('/v1/stream', baseURL)` in `stream/sse.ts` — and per the URL spec an absolute path replaces the base's path entirely. No error, a client that looks correctly configured, and every request quietly going somewhere else. There was no workaround from outside the SDK: `baseURL` was the only path input and it couldn't survive.

That pushes an ordinary deployment shape — one origin fronting several services, a BFF adding auth ahead of the data API, a Next.js/Remix/Rails route proxying through, a path-routed ingress — off the SDK and onto hand-rolled `fetch`, losing the query builder, typed errors, retry/backoff, and the streaming controller along with it.

## The fix

Took the issue's first option (relative resolution) over adding a `basePath` config field — it's what the issue title asks for and it's zero new API surface.

Request paths are now joined **onto** the base by one shared `resolveURL` helper (`clients/ts/src/url.ts`) that both transports and the codegen CLI call. Previously there were three separate URL constructions, and they disagreed: codegen's string concat already handled prefixes while the two transports didn't.

The helper normalizes the base to a directory before resolving, so a bare last segment — or a stray query/fragment shadowing one — can't eat the prefix either. A root-hosted base (`http://localhost:8080`, the overwhelmingly common case) resolves byte-identically to before; I diffed old vs new resolution across the edge cases to confirm.

## Tests

- `url.test.ts` (new) — the helper: root and prefixed bases, multi-segment prefixes, trailing-slash equivalence, params merging with a path's existing query string, base query/fragment, non-default ports.
- `stream/sse.test.ts` (new) — the SSE transport against a stubbed `EventSource`, pinning the prefix alongside the `since` and `token` params. This is the one the issue warns is easy to miss when fixing `http.ts` alone.
- `http.test.ts`, `client.test.ts` — prefix preserved through the REST transport and end-to-end from `createClient`.

## Docs

- **SDK → Creating a Client** gains a "Serving under a path prefix" section; `ClientConfig.baseURL` documents the prefix in both the table and its JSDoc. It also now states that `baseURL` must be **absolute** — a same-origin relative `/api/warehouse` throws rather than returning a `Result`, which is pre-existing behavior that was never written down and bites hardest in exactly the BFF scenario this section is about.
- **Behind a reverse proxy** gains a "Path prefixes" section — WaveHouse has no configurable base path by design, so the proxy must strip the prefix. nginx and Caddy snippets, plus a note that Kubernetes `ingress-nginx` does *not* strip by default and needs `rewrite-target`.

## Also in this PR: a `?token=` strip fix in `internal/auth`

Flagged by CodeRabbit while reviewing the docs change above, so it is worth explaining why a `fix(sdk):` PR carries a server-side auth diff.

`bearerToken` returned from the `Authorization: Bearer` branch **before** stripping `?token=` from the URL, so a request presenting both credentials left the unused JWT sitting in `r.URL` for the rest of its life. The operator-key path had the same shape one frame up, returning before `bearerToken` ran at all — exempting the single most privileged credential. The strip is now resolved once, ahead of every credential branch.

This is **defense in depth, not a fix for an observed exposure**: WaveHouse's own logging records only `r.URL.Path`, and the OTel HTTP instrumentation emits no query attribute, so nothing leaks today. It was worth closing because it was an inconsistency in an invariant the code already asserted on the query-only path, and because my own earlier commit here had just written the gap down as intended behavior — documenting it seemed strictly worse than fixing it.

Header precedence is unchanged, unrelated query params survive, and four credential combinations are pinned by tests (query-only, header+query, non-Bearer header+query, operator key+query). I verified the operator-path test genuinely catches the bug: with the fix reverted it fails with the raw JWT still in the URL.

If you would rather this land separately, it is self-contained in `internal/auth/` plus the three doc sentences that describe the behavior — say the word and I will split it out.

## Note for the reviewer

`#269` (custom headers / `fetch` override) is the sibling gap in the same config surface and is deliberately **not** in scope here.

### CODE REVIEW

File: clients/ts/src/types.ts [hunk 1]
```diff
@@ -58,7 +58,12 @@ export interface StreamSubscriber<T = Record<string, unknown>> {
 // --- Client config ---
 
 export interface ClientConfig<_DB extends Database = Database> {
-  /** Base URL of the WaveHouse server (e.g. "http://localhost:8080"). */
+  /**
+   * Base URL of the WaveHouse server (e.g. "http://localhost:8080"). May include
+   * a path prefix ("https://app.example.com/api/warehouse") for a WaveHouse
+   * behind a BFF, app-server route, or path-routed ingress; request paths are
+   * appended to it. The proxy in front must strip the prefix before forwarding.
+   */
```
[the remark this answers is not in the document]
Comment (EricAndrechek) [reply clients/ts/src/types.ts:75]: Good catch — fixed in 42c996b.

`ClientConfig.baseURL`'s JSDoc now states the absolute requirement explicitly, matching what the docs page already said:

> Must be **absolute** — scheme and host included. A relative value such as `/api/warehouse` throws a `TypeError` on the first request rather than returning a `Result`; use `` `${location.origin}/api/warehouse` ``.

Worth noting the throw is asymmetric across transports, which later review rounds tightened further (454d2ee): REST rejects, because `resolveURL` runs outside `request()`'s retry `try`, but the SSE transport catches it and reports a soft `SSE_CONNECT_ERROR` to the subscriber's *optional* `error` callback — so a stream with a bad `baseURL` can fail silently. Both paths are now documented in `sdk/reference.md`, `sdk/index.mdx`, and the SDK README.

@coderabbitai

— Claude

File: docs/src/content/docs/sdk/index.mdx [hunk 2]
```diff
@@ -323,14 +323,28 @@ const wh = createClient<Database>({
...
 | Field | Type | Default | Description |
 |-------|------|---------|-------------|
-| `baseURL` | `string` | — | WaveHouse server URL (required) |
+| `baseURL` | `string` | — | WaveHouse server URL, optionally including a path prefix (required) |
 | `auth` | `() => Promise<string> \| string` | — | Token provider. Omit for public access |
 | `options.maxRetries` | `number` | `2` | Retry attempts for failed/5xx requests |
 
 :::note[How the token is transmitted]
 The SDK attaches your `auth` token as an `Authorization: Bearer` header on REST calls, and — because the browser `EventSource` API can't set headers — as a `?token=` query parameter on streaming connections. When both are present the server reads the header in preference to the query parameter, and strips the `?token=` value from the URL after extraction so it can't leak into logs.
 :::
```
[the remark this answers is not in the document]
Comment (EricAndrechek) [reply docs/src/content/docs/sdk/index.mdx:334]: Agreed, and this was the most valuable comment on the PR — thank you. Fixed in 42c996b, then extended in f26644d.

You were right that the strip only protects WaveHouse's own logs. Two things came out of digging into it:

1. **The claim was in three places, not one.** The same "so it can't leak into logs" wording also sat in `api.md` and — worse — in `reverse-proxy.mdx`, three lines above a bullet telling the operator "does not log the full URL with `token` in it". The page contradicted itself, on exactly the page someone reads while configuring the proxy that does the logging. All three are now scoped, and I also fixed the `bearerToken` doc comment in `internal/auth/auth.go`, which was the source the docs kept copying.

2. **The corrected sentence was still wrong about *when* the strip happens.** `bearerToken` returns from the `Authorization`-header branch *before* touching `r.URL`, so when both credentials are present the query parameter is left intact — the strip only runs on the query-only path. The docs now say that.

For the record on the narrowed claim: WaveHouse's own logging only ever records `r.URL.Path`, nothing reads `r.RequestURI` (which the strip does not rewrite), and otelhttp records no query attribute — so "stays out of WaveHouse's own logs" holds, while everything upstream needs operator-side query-string redaction.

@coderabbitai

— Claude

File: clients/ts/src/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (EricAndrechek) [reply clients/ts/src/types.ts:75]: Right, and fixed in 42c996b — the JSDoc now states the absolute requirement, matching the docs page.

You and CodeRabbit landed on this independently, which was a good signal it mattered. One correction to the framing though: it is not quite "throws synchronously". `resolveURL` runs inside the async `request()`, so REST calls surface it as a **rejected promise** rather than a synchronous throw — and the SSE transport does something different again, catching it into a soft `SSE_CONNECT_ERROR` on the subscriber's optional `error` callback. The docs now describe both paths separately rather than promising one behavior for "the SDK".

— Claude

File: internal/auth/auth.go [hunk 3]
```diff
@@ -252,9 +252,12 @@ func operatorKey(r *http.Request) string {
...
 // clients that can't set headers — the ?token query parameter, which is stripped
-// from the URL so it can't leak into logs. The Authorization header takes
-// precedence when both are present; the Bearer auth-scheme is matched
-// case-insensitively (RFC 7235). Returns "" if absent.
+// from the URL so it stays out of *our own* logs (it has already crossed every
+// intermediary in the request URI, so proxies must redact query strings
+// themselves). The Authorization header takes precedence when both are present
+// — and returns before the strip, leaving the query parameter intact; the
+// Bearer auth-scheme is matched case-insensitively (RFC 7235). Returns "" if
+// absent.
```
[the remark this answers is not in the document]
Comment (EricAndrechek) [reply internal/auth/auth.go]: Agreed, and fixed in 4227b6c — plus a second instance you did not see. Thank you for this one.

**The fix.** `bearerToken` now strips `?token=` once, before either credential path returns. Header precedence and the `""`-when-absent contract are unchanged, unrelated query params survive, and requests without a `token` param never touch `params.Encode()` at all.

I checked the blast radius before changing a security-sensitive path rather than taking the diff on faith: `auth.go` is the only reader of the token query param in the codebase, no existing test pinned the old behavior (`TestMiddleware_HeaderTakesPrecedenceOverQuery` asserted only the resolved role), and every downstream handler reads through `r.URL.Query()` rather than `RawQuery`, so `Encode()`'s key reordering is invisible.

**The second instance.** Our own reviewer caught that the operator-key path has the identical shape one frame up: an `X-Operator-Key` / `Authorization: Operator` match returns from `Middleware` *before* `bearerToken` is ever called, so a break-glass request carrying `?token=<jwt>` kept the unused JWT in `r.URL` — exempting the single most privileged credential from the strip. `bearerToken(r)` is now resolved above the operator branch (aea34be), with a comment so it does not drift back down.

**Severity, for the record.** Defense in depth rather than an observed exposure: WaveHouse's own logging only ever records `r.URL.Path` (3 sites), and otelhttp v0.70.0 emits no query attribute — so nothing leaks today. What made it worth fixing is that it was an inconsistency in an invariant the code already asserted on the query-only path, and any future handler or logging change would have turned it into a real leak.

Both cases are pinned by tests — `TestMiddleware_HeaderTakesPrecedenceOverQuery` now also asserts the strip, and `TestMiddleware_OperatorKey_StripsQueryToken` is new. I verified the latter genuinely catches the bug: with the hoist reverted it fails with the raw JWT still in the URL.

@coderabbitai

— Claude

File: clients/ts/README.md [hunk 4]
```diff
@@ -125,7 +127,7 @@ const { data } = await wh.from('clicks').select('page').limit(10);
 
 ## Error Handling
 
-All operations return `{ data, error }` tuples — the SDK never throws.
+All operations return `{ data, error }` tuples — the SDK never throws for anything the server returns. It does throw on caller and environment errors: a non-absolute `baseURL` (REST calls reject with a `TypeError`; streams report `SSE_CONNECT_ERROR` to the subscriber's `error` callback; despite that error's `retryable: true`, the SDK never re-dials the stream itself), `.stream()` / `.liveQuery()` in a runtime with no `EventSource`, and an `auth` callback that rejects — a token-refresh failure propagates out of the REST call.
```
[the remark this answers is not in the document]
Comment (EricAndrechek) [reply clients/ts/README.md]: Both points are right — fixed in 4227b6c, though I wrote it differently than suggested.

`Result<T>` is an object union (`{ ok, data, error, … }`), not a tuple, and "all operations" was wrong since `.stream()` returns a `StreamController` and `.liveQuery()` a `LiveQuery`. I kept those as two sentences rather than folding them into the throws-list, which otherwise lost its subject:

> Async request methods return a `Result<T>` object — destructure it as `{ data, error }` — and never throw for anything the server returns. `.stream()` and `.liveQuery()` return controllers instead, reporting failures through the subscriber's `error` callback.

The "tuple" wording only appeared here; the docs site pages already said `Result<T>`.

Since the same absolutism turned up elsewhere once I started pulling the thread, 9b018ef also qualified `Result<T>`'s own JSDoc in `types.ts` — the canonical version, the one that shows up in an editor tooltip — which still read a flat "Never throws" while five prose locations said otherwise.

@coderabbitai

— Claude

File: clients/ts/src/types.ts [hunk 5]
```diff
@@ -58,7 +58,17 @@ export interface StreamSubscriber<T = Record<string, unknown>> {
...
+  /**
+   * Base URL of the WaveHouse server (e.g. "http://localhost:8080"). May include
+   * a path prefix ("https://app.example.com/api/warehouse") for a WaveHouse
+   * behind a backend-for-frontend (BFF), app-server route, or path-routed
+   * ingress; request paths are appended to it. The proxy in front must strip
+   * the prefix before forwarding.
+   *
+   * Must be **absolute** — scheme and host included. A relative value such as
+   * "/api/warehouse" throws a `TypeError` on the first request rather than
+   * returning a `Result`; use `` `${location.origin}/api/warehouse` ``.
```
[the remark this answers is not in the document]
Comment (EricAndrechek) [reply clients/ts/src/types.ts]: Fixed in 4227b6c — the `ClientConfig.baseURL` JSDoc now splits the two transports rather than implying they behave alike:

> The transports report a relative value such as "/api/warehouse" differently: a REST call rejects with a `TypeError` instead of returning a `Result`, while a stream reports `SSE_CONNECT_ERROR` to the subscriber's `error` callback.

Worth adding the detail that makes this matter rather than just being imprecise: the subscriber's `error` callback is **optional** (`StreamSubscriber.error?`), and `SSE_CONNECT_ERROR` carries `retryable: true` even though the SDK never re-dials a failed connect. So a stream with a relative `baseURL` can fail silently and permanently — which is the same quiet-misconfiguration failure mode #428 exists to remove. The docs now spell that out in `sdk/reference.md` alongside the two `SSE_*` codes, which were named in prose but missing from the error-code table.

@coderabbitai

— Claude
